packages/proxy/lib/http/util/service-worker-injector.ts

```diff
@@ -0,0 +1,228 @@
+        if (typeof listener === 'function') {
+          newListener = __cypressWrapListener(listener)
+        } else {
+          // since the handleEvent function could change, we need to use a proxy to wrap it
+          newListener = new Proxy(listener, {
+            get (target, key) {
+              if (key === 'handleEvent') {
+                const wrappedHandleEvent = _targetToWrappedHandleEventMap.get(target)
+                const origHandleEvent = _targetToOrigHandleEventMap.get(target)
+
+                // If the handleEvent function has not been wrapped yet, or if it has changed, we need to wrap it
+                if ((!wrappedHandleEvent && target.handleEvent) || target.handleEvent !== origHandleEvent) {
+                  _targetToWrappedHandleEventMap.set(target, __cypressWrapListener(target.handleEvent))
+                  _targetToOrigHandleEventMap.set(target, target.handleEvent)
+                }
+
+                return _targetToWrappedHandleEventMap.get(target)
+              }
+
+              return Reflect.get(target, key)
+            },
+          })
+        }
+
+        // get the capture value so we know which map to add the listener to
+        // so we can then remove the listener later if requested
+        getCaptureValue(options) ? _captureListenersMap.set(listener, newListener) : _nonCaptureListenersMap.set(listener, newListener)
+        _listenerCount++
+
+        __cypressServiceWorkerSendHasFetchEventHandlers()
+
+        return oldAddEventListener(type, newListener, options)
+      }
+
+      return oldAddEventListener(type, listener, options)
+    }
+
+    const oldRemoveEventListener = self.removeEventListener
+
+    // Overwrite the removeEventListener method so we can remove the listener from the map
+    // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener
+    self.removeEventListener = (type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions) => {
+      if (type === 'fetch' && isValidListener(listener)) {
+        // get the capture value so we know which map to remove the listener from
+        const capture = getCaptureValue(options)
+        const newListener = capture ? _captureListenersMap.get(listener) : _nonCaptureListenersMap.get(listener)
+
+        capture ? _captureListenersMap.delete(listener) : _nonCaptureListenersMap.delete(listener)
+        _listenerCount--
+
+        // If the listener is an object with a handleEvent method, we need to remove the wrapped function
+        if (typeof listener === 'object' && typeof listener.handleEvent === 'function') {
+          _targetToWrappedHandleEventMap.delete(listener)
+          _targetToOrigHandleEventMap.delete(listener)
+        }
+
+        __cypressServiceWorkerSendHasFetchEventHandlers()
+
+        return oldRemoveEventListener(type, newListener!, options)
```

Comment:
Would it be worth calling this first before our logic? That way if it throws an exception for some reason we won't inadvertently do our logic?

Reply:
good idea. [d9a876d](https://github.com/cypress-io/cypress/pull/28900/commits/d9a876ddeb1d2d9da546f506ba9fec3adebacba9)